src/lib/utils.py

```diff
@@ -425,13 +436,18 @@ def dump_manifest(contents, manifest_path):
     """Writes back 'contents' to 'manifest_path'.
 
     For YAML, we make a best-effort attempt to preserve
-    formatting; for JSON, we use the canonical 4-space indentation."""
+    formatting; for JSON, we use the canonical 4-space indentation,
+    but add a trailing newline if originally present."""
     _, ext = os.path.splitext(manifest_path)
-    with open(manifest_path, "w", encoding="utf-8") as fp:
+    with open(manifest_path, "r+", encoding="utf-8") as fp:
         if ext in (".yaml", ".yml"):
             _yaml.dump(contents, fp)
         else:
+            newline = _check_newline(fp)
             json.dump(obj=contents, fp=fp, indent=4)
+            if newline:
+                fp.write("\n")
+        fp.truncate()
```

Comment:
I'm not sure I understand why this is needed.

Reply:
I thought there might potentially be cases where the new file is shorter than the original, since we're using `r+` mode we could end up with garbage at the end of the file if we don't truncate.

Reply:
Ah, indeed, good catch. I didn't think about it when suggested to use read-write mode. Now opening opening the file twice - in read-only and write-only modes - seems like a better idea :-/

Reply:
Do you want me to go back to the original implementation?

Reply:
Yeah, if you don't mind, sorry for the confusion. Just don't change `_check_newline` argument back to a file path - instead, open the file for reading and detecting trailing newline here, in `dump_manifest`.

Reply:
So if I understand this right, you want me to keep `_check_newline` as it currently is but just change `dump_manifest` back to `w` mode and open it for reading beforehand? I'll send a new commit and see if that's what you want.

Reply:
Yep, something like this.